### PR TITLE
Bug Fixes on some paths of the native side

### PR DESCRIPTION
**Pull request by Anipik**
Fixes https://github.com/dotnet/machinelearning/issues/1344
These bugs were not caught by the tests earlier because everything is still sending the stuff as Aligned. The code changes are made to just match with managed side code. no new logic has been added here

cc @eerhardt  @yaeldekel @tannergooding

### CODE REVIEW

File: src/Microsoft.ML.CpuMath/SseIntrinsics.cs [hunk 1]
```diff
@@ -221,7 +221,7 @@ public static unsafe void MatMul(float[] mat, float[] src, float[] dst, int crow
                             length -= misalignment;
                         }
 
-                        if (length > 4)
+                        if (length > 3)
```
Comment (Ivanidzo4ka) [inline src/Microsoft.ML.CpuMath/SseIntrinsics.cs]: >> 3 [](start = 35, length = 3)

Not sure about performance, but readability wise, i would change it to >=4 instead of >3.

File: src/Native/CpuMathNative/Sse.cpp [hunk 2]
```diff
@@ -194,7 +194,7 @@ EXPORT_API(void) MatMul(_In_ const float * pmat, _In_ const float * psrc, _Inout
             if (length > 3)
             {
                 remainder = length % 4;
-                while(pSrcCurrent < pSrcEnd)
+                while(pSrcCurrent + 4 <= pSrcEnd)
```
Comment (Ivanidzo4ka) [inline src/Native/CpuMathNative/Sse.cpp]: >pSrcCurrent [](start = 22, length = 11)

A) can you run auto format on this file, I can see bunch of empty spaces after line.
B) more importantly, can you add comments to code? I assume you figured out what is going on here, would be nice to have some comments on top of if statements, overwise I would probably spend half of hour by just looking into all this conditions before I get to code you changed.

File: src/Microsoft.ML.CpuMath/SseIntrinsics.cs (hunk 1 above)
Comment (Anipik) [reply src/Microsoft.ML.CpuMath/SseIntrinsics.cs]: cc @tannergooding

Comment (tannergooding) [reply src/Microsoft.ML.CpuMath/SseIntrinsics.cs]: I'm not sure I agree from a readability standpoint, most code I see operates in terms of `<` and `>` by default and `length > 3` is generally easy to read/understand.

**Comment by Anipik** [comment]
> Are there existing tests that are failing because of this? Or a new test is needed? 

The existing tests for netstandard are just checking the alignedcodePath i.e the existing tests are not failing. I was able to caught the errors when I remove the aligned code path from netstandard and we were no longer ensuring aligned input for netstandard.

The tests maybe enough after if we have removed aligned array but it requires more investigation that we are not missing any corner case.

For this PR,  i will recommend to add just this fixs as they make the tests pass without alignment too in netstandard

**Comment by Anipik** [comment]
@tannergooding I break the commit into 2 commits

File: src/Native/CpuMathNative/Sse.cpp [hunk 3]
```diff
@@ -43,1299 +43,1299 @@
 
 const unsigned int LeadingAlignmentMask[16] =
 {
-    0x00000000, 0x00000000, 0x00000000, 0x00000000,
-    0xFFFFFFFF, 0x00000000, 0x00000000, 0x00000000,
-    0xFFFFFFFF, 0xFFFFFFFF, 0x00000000, 0x00000000,
-    0xFFFFFFFF, 0xFFFFFFFF, 0xFFFFFFFF, 0x00000000,
+	0x00000000, 0x00000000, 0x00000000, 0x00000000,
```
Comment (eerhardt) [inline src/Native/CpuMathNative/Sse.cpp]: Why are we adding tabs here? We use spaces everywhere else.

Comment (Anipik) [reply src/Native/CpuMathNative/Sse.cpp]: autoformat for .cpp somehow used tabs for autoformatting. I will correct it

File: src/Native/CpuMathNative/Sse.cpp [hunk 4]
```diff
@@ -913,9 +945,9 @@ EXPORT_API(float) Sum(const float* pValues, int length)
 
         switch (length)
         {
-            case 3: result += pValues[2];
-            case 2: result += pValues[1];
-            case 1: result += pValues[0];
+        case 3: result += pValues[2];
```
Comment (eerhardt) [inline src/Native/CpuMathNative/Sse.cpp]: I think the case statements should be indented.
